File: 计算机工程实践/Transbot SE/pi/src/catkin_ws/src/learning_topic/src/turtle_pose_subscriber.cpp

/*创建一个小海龟的当前位姿信息接收*/
#include <ros/ros.h>
#include "turtlesim/Pose.h"
// 接收消息后，会进入消息回调函数，回调函数里边会对接收到的数据进行处理
void turtle_poseCallback(const turtlesim::Pose::ConstPtr& msg){
    // 打印接收到的消息
    ROS_INFO("Turtle pose: x:%0.3f, y:%0.3f", msg->x, msg->y);
}

int main(int argc, char **argv){
    
    ros::init(argc, argv, "turtle_pose_subscriber");// 初始化ROS节点

    ros::NodeHandle n;//这里是创建句柄

    // 创建一个订阅者，订阅的话题是/turtle1/pose的topic，poseCallback是回调函数
    ros::Subscriber pose_sub = n.subscribe("/turtle1/pose", 10, turtle_poseCallback);
   
    ros::spin(); // 循环等待回调函数

    return 0;
}

File: 计算机工程实践/Transbot SE/pi/src/catkin_ws/src/learning_topic/scripts/turtle_pose_subscriber.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import rospy
from turtlesim.msg import Pose

def poseCallback(msg):
    rospy.loginfo("Turtle pose: x:%0.3f, y:%0.3f", msg.x, msg.y)

def turtle_pose_subscriber():
    
    rospy.init_node('turtle_pose_subscriber', anonymous=True)# ROS节点初始化

    # 创建一个Subscriber，订阅名为/turtle1/pose的topic，注册回调函数poseCallback
    rospy.Subscriber("/turtle1/pose", Pose, poseCallback)

    
    rospy.spin()# 循环等待回调函数

if __name__ == '__main__':
    turtle_pose_subscriber()

